pkg/csi/service/vanilla/controller.go

```diff
@@ -2265,11 +2266,15 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 			nodevm, err = c.nodeMgr.GetNodeByName(ctx, req.NodeId)
 		}
 		if err != nil {
-			if err == cnsvsphere.ErrVMNotFound {
-				log.Infof("Virtual Machine for Node ID: %v is not present in the VC Inventory. "+
-					"Marking ControllerUnpublishVolume for Volume: %q as successful.", req.NodeId, req.VolumeId)
-				return &csi.ControllerUnpublishVolumeResponse{}, "", nil
-			} else {
+			if err == cnsnode.ErrNodeNotFound {
```

Comment:
When we get `cnsvsphere.ErrVMNotFound ` at line 2268, it is already confirmed that VM is not present in the VC inventory. Why do we need to check it again using DNS name?

Reply:
when UseCSINodeId feature is enabled, when get can call to detach the volume we will get Node VM UUID in the controllerunpublish request. Using Node VM UUID we can look up if Node VM is present on the system or not. if it is present we will get VM MoRef and using which we will issue detach call on the vCenter.

Using DNS to look up a node VM may not be a reliable method, especially if there are multiple VMs with the same DNS name on the vCenter. A better option is to use the UUID of the VM as a unique identifier to look up the node VM on the vCenter. The UUID is a unique identifier for each VM, making it the best option for verifying the presence of a specific node VM on the vCenter. Therefore, it is recommended to use the UUID of the VM instead of DNS when looking up a node VM.

Reply:
Regarding the first issue, the transitive closure of calls under `c.nodeMgr.GetNodeByName()` may return errors for a large number of reasons outside of node failure. The code should not respond that the volume has been released unless it is confirmed that the node was actually lost. 

I will update the patch with the UUID changes, thank you for that feedback!

Reply:
Regarding the use of DNS vs. node UUID, this code has been in production for some time and I forgot about the mechanics mentioned in line 2271: "As the VM UUID is not known anymore, need to search by using NodeID as DNS Name". Do you have some consideration of this state?

Reply:
if Node can be returned using `c.nodeMgr.GetNodeByName`, because Node object is deleted from API server. We do deeper look into vCenter using `c.nodeMgr.GetNodeByUuid(ctx, req.NodeId)`.

Refer to following implementation and let us know what is missing in checking Node VM existence before we mark detach as scuess. 

```golang
nodevm, err = c.nodeMgr.GetNodeByName(ctx, req.NodeId)
			if err == node.ErrNodeNotFound {
				log.Infof("Performing node VM lookup using node VM UUID: %q", req.NodeId)
				nodevm, err = c.nodeMgr.GetNodeByUuid(ctx, req.NodeId)
			}
```


```golang
func (nodes *Nodes) GetNodeByUuid(ctx context.Context, nodeUuid string) (*cnsvsphere.VirtualMachine, error) {
	return nodes.cnsNodeManager.GetNode(ctx, nodeUuid, nil)
}
```


```golang
func (m *defaultManager) GetNode(ctx context.Context,
	nodeUUID string, dc *vsphere.Datacenter) (*vsphere.VirtualMachine, error) {
	log := logger.GetLogger(ctx)
	vmInf, discovered := m.nodeVMs.Load(nodeUUID)
	if !discovered {
		log.Infof("Node hasn't been discovered yet with nodeUUID %s", nodeUUID)
		var vm *vsphere.VirtualMachine
		var err error
		if dc != nil {
			vm, err = dc.GetVirtualMachineByUUID(ctx, nodeUUID, false)
			if err != nil {
				log.Errorf("failed to find node with nodeUUID %s on datacenter: %+v with err: %v", nodeUUID, dc, err)
				return nil, err
			}
			m.nodeVMs.Store(nodeUUID, vm)
		} else {
			if err = m.DiscoverNode(ctx, nodeUUID); err != nil {
				log.Errorf("failed to discover node with nodeUUID %s with err: %v", nodeUUID, err)
				return nil, err
			}

			vmInf, _ = m.nodeVMs.Load(nodeUUID)
			vm = vmInf.(*vsphere.VirtualMachine)
		}
		log.Infof("Node was successfully discovered with nodeUUID %s in vm %v", nodeUUID, vm)
		return vm, nil
	}

	vm := vmInf.(*vsphere.VirtualMachine)
	log.Debugf("Renewing virtual machine %v with nodeUUID %q", vm, nodeUUID)

	if err := vm.Renew(ctx, true); err != nil {
		log.Errorf("failed to renew VM %v with nodeUUID %q with err: %v", vm, nodeUUID, err)
		return nil, err
	}

	log.Debugf("VM %v was successfully renewed with nodeUUID %q", vm, nodeUUID)
	return vm, nil
}
```

Reply:
Thanks @divyenpatel I misread your initial response. Understand what you are after now and will get that resolved. Switching this to draft.